instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/WaypointsTest.kt

```diff
@@ -0,0 +1,256 @@
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .routingTilesOptions(
+                        RoutingTilesOptions.Builder()
+                            .tilesBaseUri(URI(mockWebServerRule.baseUrl))
+                            .build()
+                    )
+                    .navigatorPredictionMillis(0L)
+                    .build()
+            )
+            mockWebServerRule.requestHandlers.clear()
+        }
+    }
+
+    @Test
+    fun ev_route_with_waypoints_in_response_root_by_default() = sdkTest {
+        addResponseHandler(R.raw.ev_route_response_with_waypoints_in_root, evCoordinates)
+        val routes = requestRoutes(evCoordinates, electric = true, waypointsPerRoute = null)
+
+        checkWaypointsInRoot(expectedEvWaypointsNamesAndLocations, routes[0])
+    }
+
+    @Test
+    fun ev_route_with_waypoints_in_response_root() = sdkTest {
+        addResponseHandler(R.raw.ev_route_response_with_waypoints_in_root, evCoordinates)
+        val routes = requestRoutes(evCoordinates, electric = true, waypointsPerRoute = false)
+
+        checkWaypointsInRoot(expectedEvWaypointsNamesAndLocations, routes[0])
+    }
+
+    @Test
+    fun ev_route_with_waypoints_per_route() = sdkTest {
+        addResponseHandler(R.raw.ev_route_response_with_waypoints_per_route, evCoordinates)
+        val routes = requestRoutes(evCoordinates, electric = true, waypointsPerRoute = true)
+
+        checkWaypointsPerRoute(expectedEvWaypointsNamesAndLocations, routes[0])
+    }
+
+    @Test
+    fun non_ev_route_with_waypoints_in_response_root_by_default() = sdkTest {
+        addResponseHandler(R.raw.route_response_with_waypoints_in_root, nonEvCoordinates)
+        val routes = requestRoutes(nonEvCoordinates, electric = false, waypointsPerRoute = null)
+
+        checkWaypointsInRoot(expectedFirstNonEvWaypointsNamesAndLocations, routes[0])
+        checkWaypointsInRoot(expectedFirstNonEvWaypointsNamesAndLocations, routes[1])
+    }
+
+    @Test
+    fun non_ev_route_with_waypoints_in_response_root() = sdkTest {
+        addResponseHandler(R.raw.route_response_with_waypoints_in_root, nonEvCoordinates)
+        val routes = requestRoutes(nonEvCoordinates, electric = false, waypointsPerRoute = false)
+
+        checkWaypointsInRoot(expectedFirstNonEvWaypointsNamesAndLocations, routes[0])
+        checkWaypointsInRoot(expectedFirstNonEvWaypointsNamesAndLocations, routes[1])
+    }
+
+    @Test
+    fun non_ev_route_with_waypoints_per_route() = sdkTest {
+        addResponseHandler(R.raw.route_response_with_waypoints_per_route, nonEvCoordinates)
+        val routes = requestRoutes(nonEvCoordinates, electric = false, waypointsPerRoute = true)
+
+        checkWaypointsPerRoute(expectedFirstNonEvWaypointsNamesAndLocations, routes[0])
+        checkWaypointsPerRoute(expectedSecondNonEvWaypointsNamesAndLocations, routes[1])
+    }
```

Comment:
They seem like unit tests which are executed on device. I understand why it's done like this, we can't access NN in unit tests. I added NAVSDK-726 to address absence of NN in unit tests later.

Reply:
No, not because of that. It's not a unit test. It's an integration test: we request routes and parse the response.
We test the significant part of the SDK, not some specific class.
We already have such tests: see `NavigationRouteTest`.
And we don't have access to NN classes here either: everything that's checked here is represented by the SDK classes.

Reply:
It seems that `NavigationRoute#internalWaypoints` is calculated using NN, isn't it?